homeassistant/components/met/__init__.py

```diff
@@ -1,24 +1,155 @@
 """The met component."""
+from datetime import timedelta
+import logging
+from random import randrange
+
+import metno
+
+from homeassistant.const import (
+    CONF_ELEVATION,
+    CONF_LATITUDE,
+    CONF_LONGITUDE,
+    EVENT_CORE_CONFIG_UPDATE,
+    LENGTH_FEET,
+    LENGTH_METERS,
+)
 from homeassistant.core import Config, HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util.distance import convert as convert_distance
+import homeassistant.util.dt as dt_util
+
+from .const import CONF_TRACK_HOME, DOMAIN
+
+URL = "https://api.met.no/weatherapi/locationforecast/2.0/classic"
 
-from .config_flow import MetFlowHandler  # noqa: F401
-from .const import DOMAIN  # noqa: F401
+
+_LOGGER = logging.getLogger(__name__)
 
 
 async def async_setup(hass: HomeAssistant, config: Config) -> bool:
     """Set up configured Met."""
+    hass.data.setdefault(DOMAIN, {})
     return True
 
 
 async def async_setup_entry(hass, config_entry):
     """Set up Met as config entry."""
+    coordinator = MetDataUpdateCoordinator(hass, config_entry)
+    await coordinator.async_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    if config_entry.data.get(CONF_TRACK_HOME, False):
+        coordinator.track_home()
+
+    hass.data[DOMAIN][config_entry.entry_id] = coordinator
+
     hass.async_create_task(
         hass.config_entries.async_forward_entry_setup(config_entry, "weather")
     )
+
     return True
 
 
 async def async_unload_entry(hass, config_entry):
     """Unload a config entry."""
     await hass.config_entries.async_forward_entry_unload(config_entry, "weather")
+    hass.data[DOMAIN][config_entry.entry_id].untrack_home()
+    hass.data[DOMAIN].pop(config_entry.entry_id)
+
     return True
+
+
+class MetDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching Met data."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize global Met data updater."""
+        self._unsub_track_home = None
+        self.weather = MetWeatherData(
+            hass, config_entry.data, hass.config.units.is_metric
+        )
+        self.weather.init_data()
+
+        update_interval = timedelta(minutes=randrange(55, 65))
```

Comment:
Codacy found an issue: [Standard pseudo-random generators are not suitable for security/cryptographic purposes.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)